snapshots/wire-1.0.yaml

```diff
@@ -0,0 +1,118 @@
+# A custom Stack snapshot (https://docs.haskellstack.org/en/stable/custom_snapshot/) used for
+# Wire code. Should be never changed once pushed to develop.
```

Comment:
what happens if it does?  how can i recover?  (trying to understand, i'm not suggesting we should ever do that.)

Reply:
Stack has two ways to refer to snapshots:

* URL
* filepath

In wire-server we refer to the snapshot by filepath. If the snapshot file changes, Stack will detect changes and treat it as a new snapshot (all snapshots are indexed by hash so if you change the file and then change it back, the only thing you should do is delete the second snapshot from your system).

In other repositories we will refer to snapshots by URL. Stack only downloads URLs once and never checks them again. So, when building code from other repositories, people would be getting inconsistent results depending on whether they built that code before or not.

Reply:
Typo:

```suggestion
# Wire code. Should never be changed once pushed to develop.
```

Do I understand correctly that if we want to bump the version of a library, `snapshots/wire-1.1.yaml` (a copy of this file with one version changed) file would be created, and `stack.yaml` would from then on refer to wire-1.1?

> Should be never changed once pushed to develop.

The snapshot docs specify "custom snapshot[s] are assumed to be mutable". So is this a guideline for us only? Perhaps add a comment on how to deal with added, removed, or updated dependencies.

Reply:
> Do I understand correctly that if we want to bump the version of a library, snapshots/wire-1.1.yaml (a copy of this file with one version changed) file would be created, and stack.yaml would from then on refer to wire-1.1?

Not a copy, more like a patch. Snapshots can be based on other snapshots. So, it would be something like:

```yaml
resolver: <link to wire-1.0.yaml>

packages: [updated packages here]

# there is also an ability to remove packages but I don't remember how it's done
```

Actually if we want that to be usable in other repositories, we should use a link in our stack.yaml as well, not a filepath. I'll amend the PR.

> The snapshot docs specify "custom snapshot[s] are assumed to be mutable".

If you refer to snapshots by filepath, then yes, they can be mutable. Stack will notice changes and just create a new snapshot (leaving the old one available, too).

If you refer to snapshots by URL, as we're planning to do in other repos, then they have to be immutable because Stack is only going to download the URL once.